conda/core/solve.py

```diff
@@ -1324,28 +1345,65 @@ def _configure_solver_for_remove(self, state):
-        specs_to_add = list(self.specs_to_add)
+        task = (api.SOLVER_UPDATE if self._command == "update" else api.SOLVER_INSTALL)
+        task_and_specs[task].update([s.conda_build_form() for s in self.specs_to_add])
+
         # B. Implicitly requested via `update --all`: add everything in prefix
         if self._command == "update" and update_modifier == UpdateModifier.UPDATE_ALL:
-            for pkg in state["installed_pkgs"]:
-                if not pkg.is_unmanageable:  # skip virtual packages
-                    specs_to_add.append(MatchSpec(pkg.name))
-        # C. Implicitly requested via `install|update --update-deps`
+            task_and_specs[api.SOLVER_UPDATE].update(
+                [
+                    MatchSpec(pkg.name).conda_build_form()
+                    for pkg in state["installed_pkgs"]
+                    if not pkg.is_unmanageable
+                ]
+            )
+
+        # C. Implicitly requested via `install --update-deps`
         elif update_modifier == UpdateModifier.UPDATE_DEPS:
-            final_specs = specs_to_add
-            for spec in specs_to_add:
-                prec = next(p for p in state["installed_pkgs"] if p.name == spec.name)
-                for dep in prec.depends:
-                    ms = MatchSpec(dep)
-                    if ms.name != "python":
-                        final_specs.append(MatchSpec(ms.name))
-            specs_to_add += list(set(final_specs))
-        return [s.conda_build_form() for s in specs_to_add]
+            # This code below is adapted from the legacy solver logic
+            # found in Solver._post_sat_handling()
+
+            # We need a pre-solve to find the full chain of dependencies
+            # for the requested specs
+            solved_pkgs = self.solve_final_state(
+                update_modifier=UpdateModifier.UPDATE_SPECS,
+                deps_modifier=deps_modifier,
+                ignore_pinned=ignore_pinned,
+                force_remove=force_remove,
+                force_reinstall=force_reinstall)
+            graph = PrefixGraph(solved_pkgs)
+            update_names = set()
+            for spec in self.specs_to_add:
+                node = graph.get_node_by_name(spec.name)
+                update_names.update(ancest_rec.name for ancest_rec in graph.all_ancestors(node))
+            specs_map = {name: MatchSpec(name) for name in update_names}
+
+            # Remove pinned_specs and any python spec (due to major-minor pinning business rule).
+            # Add in the original specs_to_add on top.
+            for spec in self._pinned_specs():
+                specs_map.pop(spec.name, None)
+            if "python" in specs_map:
+                python_rec = next((p for p in state["installed_pkgs"] if p.name == "python"), None)
+                if python_rec is not None:
+                    py_ver = ".".join(python_rec.version.split(".")[:2]) + ".*"
```

Comment:
This seems a bit brittle, is there a chance that `python_rec.version` could be an empty string or nonsensible version?